types/wlr_input_method_v2.c

```diff
@@ -108,10 +116,126 @@ static void im_delete_surrounding_text(struct wl_client *client,
+	struct wlr_input_popup_surface_v2 *popup_surface = surface->role_data;
+	if (popup_surface == NULL) {
+		return;
+	}
+	if (surface->pending.committed & WLR_SURFACE_STATE_BUFFER &&
+			surface->pending.buffer_resource == NULL) {
+		// This is a NULL commit
+		popup_surface_set_mapped(popup_surface, false);
+	}
+}
+
+static void popup_surface_surface_role_commit(struct wlr_surface *surface) {
+	struct wlr_input_popup_surface_v2 *popup_surface = surface->role_data;
+	if (popup_surface == NULL) {
+		return;
+	}
+	popup_surface_set_mapped(popup_surface, wlr_surface_has_buffer(surface)
+		&& popup_surface->input_method->client_active);
+}
+
+const struct wlr_surface_role input_popup_surface_v2_role = {
+	.name = "input_popup",
+	.precommit = popup_surface_surface_role_precommit,
+	.commit = popup_surface_surface_role_commit,
+};
+
+static void popup_surface_destroy(
+		struct wlr_input_popup_surface_v2 *popup_surface) {
+	popup_surface_set_mapped(popup_surface, false);
+	wlr_signal_emit_safe(&popup_surface->events.destroy, popup_surface);
+	wl_list_remove(&popup_surface->surface_destroy.link);
+	wl_list_remove(&popup_surface->link);
+	free(popup_surface);
+}
+
+static void popup_surface_handle_surface_destroy(struct wl_listener *listener,
+		void *data) {
+	struct wlr_input_popup_surface_v2 *popup_surface =
+		wl_container_of(listener, popup_surface, surface_destroy);
+	popup_surface_destroy(popup_surface);
+}
+
 static void im_get_input_popup_surface(struct wl_client *client,
 		struct wl_resource *resource, uint32_t id,
 		struct wl_resource *surface) {
-	wlr_log(WLR_INFO, "Stub: zwp_input_method_v2::get_input_popup_surface");
+	struct wlr_input_method_v2 *input_method =
+		input_method_from_resource(resource);
+	if (!input_method) {
+		return;
+	}
+
+	struct wl_resource *popup_resource = wl_resource_create(
+		client, &zwp_input_popup_surface_v2_interface,
+		wl_resource_get_version(resource), id);
+	if (!popup_resource) {
+		wl_client_post_no_memory(client);
+		return;
+	}
+
+	struct wlr_input_popup_surface_v2 *popup_surface =
+		calloc(1, sizeof(struct wlr_input_popup_surface_v2));
+	if (!popup_surface) {
+		wl_client_post_no_memory(client);
+		return;
+	}
+
+	struct wlr_surface *wlr_surface = wlr_surface_from_resource(surface);
+	if (!wlr_surface_set_role(wlr_surface, &input_popup_surface_v2_role,
+			popup_surface, resource, ZWP_INPUT_METHOD_V2_ERROR_ROLE)) {
+		free(popup_surface);
+		return;
+	}
+
+	popup_surface->resource = popup_resource;
+	popup_surface->input_method = input_method;
+	popup_surface->surface = wlr_surface;
+	wl_signal_add(&popup_surface->surface->events.destroy,
+		&popup_surface->surface_destroy);
+	popup_surface->surface_destroy.notify =
+		popup_surface_handle_surface_destroy;
+
+	wl_signal_init(&popup_surface->events.map);
+	wl_signal_init(&popup_surface->events.unmap);
+	wl_signal_init(&popup_surface->events.destroy);
+
+	popup_surface_set_mapped(popup_surface,
```

Comment:
Should we allow already-mapped surfaces? I guess the protocol allows them, and disallowing them would be a breaking change. But before getting the text input box, the client can't really know how to draw the popup.